sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotExposeListT.cs

```diff
@@ -0,0 +1,150 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotExposeListT : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3956";
+        private const string MessageFormat = "Refactor this {0} to use a generic collection designed for inheritance.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (c.IsTest())
+                {
+                    return;
+                }
+
+                var baseMethodDeclaration = (BaseMethodDeclarationSyntax)c.Node;
+                var methodSymbol = c.SemanticModel.GetDeclaredSymbol(baseMethodDeclaration);
+
+                if (methodSymbol == null ||
+                    !IsPubliclyAccessible(methodSymbol) ||
+                    methodSymbol.IsOverride ||
+                    IsPropertyGetOrSet(methodSymbol))
```

Comment:
It could be few different things as per MethodKind. I'll change that to only check for methods and constructors.